pyminiaudio/miniaudio.py

```diff
@@ -894,72 +975,94 @@ def __init__(self, ma_output_format: int = ma_format_s16, nchannels: int = 2,
+    def start(self, callback_generator: PlaybackCallbackGeneratorType) -> None:
+        """Start the audio device: playback begins. The audio data is provided by the given callback generator.
         The generator gets sent the required number of frames and should yield the sample data
         as raw bytes or as an array.array.  (it should already be started before passing it in)"""
-        if self.audio_producer:
-            raise MiniaudioError("can't start an already started device")
-        if not inspect.isgenerator(audio_producer):
-            raise TypeError("audio producer must be a generator", type(audio_producer))
-        self.audio_producer = audio_producer
-        result = lib.ma_device_start(self._device)
-        if result != lib.MA_SUCCESS:
-            raise MiniaudioError("failed to start audio device", result)
-
-    def stop(self) -> None:
-        """Halt playback."""
-        self.audio_producer = None
-        result = lib.ma_device_stop(self._device)
-        if result != lib.MA_SUCCESS:
-            raise MiniaudioError("failed to stop audio device", result)
-
-    def close(self):
-        """Halt playback and close down the device."""
-        self.audio_producer = None
-        if self._device is not None:
-            lib.ma_device_uninit(self._device)
-            self._device = None
-        if id(self) in _callback_data:
-            del _callback_data[id(self)]
+        return super().start(callback_generator)
 
     def data_callback(self, device: ffi.CData, output: ffi.CData, input: ffi.CData, framecount: int) -> None:
-        if self.audio_producer:
+        if self.callback_generator:
             try:
-                samples = self.audio_producer.send(framecount)
+                samples = self.callback_generator.send(framecount)
             except StopIteration:
-                self.audio_producer = None
+                self.callback_generator = None
                 return
             except Exception:
-                self.audio_producer = None
+                self.callback_generator = None
                 raise
             samples_bytes = _bytes_from_generator_samples(samples)
             if samples_bytes:
                 if len(samples_bytes) > framecount * self.sample_width * self.nchannels:
-                    self.audio_producer = None
+                    self.callback_generator = None
                     raise MiniaudioError("number of frames from callback exceeds maximum")
                 ffi.memmove(output, samples_bytes, len(samples_bytes))
 
 
+class DuplexStream(AbstractDevice):
+    def __init__(self, playback_format: int = ma_format_s16, playback_channels: int = 2, capture_format: int = ma_format_s16, capture_channels: int = 2, sample_rate: int = 44100, buffersize_msec: int = 200, playback_device_id: Union[ffi.CData, None] = None, capture_device_id: Union[ffi.CData, None] = None):
```

Comment:
Line too long (please run ``pycodestyle`` tool to check this)
also missing   -> None